Session management + user properties + client to send events

eliasyishak opened the pull request:
This PR contains all the action items from issue: https://github.com/eliasyishak/dash_analytics/issues/14

A session management class has been created that is responsible for extending the session's `last_ping` variable if we are within the session duration – if it is not, we will create a new session id.

A user properties class has also been added to make it easier to pass information about the user and their machine. It also comes with its own methods that will prepare the data it contains for the json payload that will get sent to the GA4 endpoint for data collection.

A GA client class has also been added to handle sending the POST request itself with the json payload for data collection

-------

General updates for this PR:

`analytics.dart`
- added platform as an arg for the class
- using the `constants.dart` file to store constant values that are getting passed 
- added a close method which will close the http client in `GAClient`
- added the sendEvent method to take in `eventName` and `eventData` and use `GAClient` to send data
- instead of default values in the `AnalyticsImpl` class, we will use the factory constructor in the abstract class to add the default values (allows for easier testing)
- using the `GAClient` class within the `AnalyticsImpl` class to send data
- creating a `UserProperty` instance in the constructor for `AnalyticsImpl` to store metadata around the user

`config_handler.dart`
- using `package:clock` for date stamps which allows for easier testing when checking session length
- using the `constants.dart` file to store the directory and file names for the files generated

`ga_client.dart`
- created this class to send events to Google Analytics to abstract away creating the `HttpClient` from the main `Analytics` class

`initializer.dart`
- moving the functions that created the initial files into the `Initializer` class instead of being standalone functions
- using the constants from `constants.dart` to reference the filenames that need to be created

`session.dart`
- new class that will handle session management on the user's computer
- a session will be at minimum 30 minutes (or whatever is defined in `constants.dart`), as the user sends events, the session will end 30 minutes after the latest event sent

`user_property.dart`
- this class will hold all the metadata around a user and will manage getting the latest session id from the `Session` class (extends the session or creates a new session accordingly)

eliasyishak committed:
injecting a clock to make it easier to write tests for session rolling

eliasyishak committed:
renaming test file

eliasyishak committed:
refactor to use constants for filenames

eliasyishak committed:
refactor to explicitly pass args to [AnalyticsImpl]

eliasyishak committed:
added GAClient + Session classes

eliasyishak commented:
I have also implemented a `constants.dart` file to define all the constant values that get used within the package such as the tools message we will show the user, the duration for a session, etc.

eliasyishak committed:
various updates - details in desc

- adding the platform as something that gets passed to the class to not depend on dart:io
- the analytics class sendevent method uses arguments to collect the data that gets sent
- using the user property class to manage the user metadata

eliasyishak commented:
General updates for this PR:

`analytics.dart`
- added platform as an arg for the class
- using the constants.dart file to store constant values that are getting passed 
- added a close method which will close the http client in GAClient
- added the sendEvent method to take in eventName and eventData and use GAClient to send data
- instead of default values in the AnalyticsImpl class, we will use the factory constructor in the abstract class to add the default values (allows for easier testing)
- using the GAClient class within the AnalyticsImpl class to send data
- creating a UserProperty instance in the constructor for AnalyticsImpl to store metadata around the user

`config_handler.dart`
- using `package:clock` for date stamps which allows for easier testing when checking session length
- using the `constants.dart` file to store the directory and file names for the files generated

`ga_client.dart`
- created this class to send events to Google Analytics to abstract away creating the HttpClient from the main `Analytics` class

`initializer.dart`
- moving the functions that created the initial files into the `Initializer` class instead of being standalone functions
- using the constants from `constants.dart` to reference the filenames that need to be created

`session.dart`
- new class that will handle session management on the user's computer
- a session will be at minimum 30 minutes (or whatever is defined in `constants.dart`), as the user sends events, the session will end 30 minutes after the latest event sent

`user_property.dart`
- this class will hold all the metadata around a user and will manage getting the latest session id from the `Session` class (extends the session or creates a new session accordingly)

eliasyishak commented on example/dash_analytics_example.dart:
```diff
@@ -40,13 +38,19 @@ final Analytics analytics = Analytics(
   branch: 'ey-test-branch',
   flutterVersion: 'Flutter 3.6.0-7.0.pre.47',
   dartVersion: 'Dart 2.19.0',
+  platform: 'macos',
```
Make this an enum to make sure we have no variations in the strings

eliasyishak commented on example/dash_analytics_example.dart:
```diff
@@ -40,13 +38,19 @@ final Analytics analytics = Analytics(
 
   print(analytics.telemetryEnabled);
+  analytics.sendEvent(
+    eventName: 'testing_from_dash',
```
Look into Dart's enhanced enums for keeping event names

christopherfujino replied:
https://dart.dev/guides/language/language-tour#enumerated-types

eliasyishak replied:
Thank you! This is perfect, snippet of sample code

```dart
enum Events {
  // List all of the events here
  runTime('run_time'),
  hotReloadTime('hot_reload_time');

  // The positional arg for the default constructor are the labels we have
  // defined in the values above
  const Events(this.label);
  final String label;
}

void main() {
  const event = Events.runTime;
  print(event.label); // Returns 'run_time'
}
```

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -26,6 +31,10 @@ abstract class Analytics {
+        platform: platform,
+        toolsMessage: kToolsMessage,
+        toolsMessageVersion: kToolsMessageVersion,
+        fs: LocalFileSystem(),
```
I think this can/should be const
```suggestion
        fs: const LocalFileSystem(),
```

eliasyishak committed:
Update analytics.dart

christopherfujino commented on lib/src/config_handler.dart:
```diff
@@ -50,8 +52,8 @@ class ConfigHandler {
           homeDirectory.path,
-          '.dart-tool',
-          'dart-flutter-telemetry.config',
+          kDartToolDirectoryName,
```
Should this be `'.dart'`?

eliasyishak replied:
I wasn't sure on the naming for the constants which is why i created the `constants.dart` file so we can change on the fly.

The PDD did mention a `.dart-tool` directory though and the design doc seems to suggest something else...

christopherfujino replied:
I'm not too opinionated on this. I just suggested `.dart` because I already have a user-home level `.dart` but not `.dart-tool`, and I thought `.dart-tool` was specifically for projects.

eliasyishak replied:
That is a good point, I will relabel this constant

eliasyishak committed:
creating test analytics class which removes sendEvent method

eliasyishak committed:
adding documentation for json schema for ga4

eliasyishak committed:
tests added to check that sessions are being managed correctly

eliasyishak committed:
using analytics method for getting session id instead of reading file

eliasyishak committed:
adding enum type for platform

eliasyishak committed:
using enums for eventName and platform

christopherfujino commented on example/dash_analytics_example.dart:
```diff
@@ -4,13 +4,12 @@ import 'package:file/file.dart';
 import 'package:file/local.dart';
 
 import 'package:dash_analytics/dash_analytics.dart';
+import 'package:dash_analytics/src/enums.dart';
```
package consumers generally shouldn't import files under `lib/src` (there's an analyzer lint for this, but apparently it's not part of package:lints?) (note, it's ok for tests to directly import any file). Instead, can you make `package:dash_analytics/dash_analytics.dart` export everything that users would need to directly reference?

christopherfujino replied:
This is an example of dart file at the top-level lib directory that exists to control what from lib/src will be exported to end users: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/foundation.dart

eliasyishak replied:
Oh whoops, I completely forgot to replace that – the auto import is too good 😄

eliasyishak committed:
export enums from `lib/dash_analytics.dart`

christopherfujino commented on example/dash_analytics_example.dart:
```diff
@@ -35,18 +34,23 @@ final Analytics analytics = Analytics(
   branch: 'ey-test-branch',
   flutterVersion: 'Flutter 3.6.0-7.0.pre.47',
   dartVersion: 'Dart 2.19.0',
+  platform: DevicePlatforms.macos,
```
This shouldn't be hard-coded, right?

eliasyishak replied:
I had originally just hard coded this in because it was for the example code, ideally each client would have their own logic for determining the platform and they would pass it in

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -69,15 +81,35 @@ abstract class Analytics {
+  /// Returns a map representation of the [UserProperty] for the [Analytics] instance
+  ///
+  /// This is what will get sent to Google Analytics with every request
+  Map get userPropertyMap;
```
Provide explicit types

```suggestion
  Map<String, String> get userPropertyMap;
```

christopherfujino replied:
Can you also enable the `always_specify_types` and `always_declare_return_types` lints in analysis_options.yaml?

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -69,15 +81,35 @@ abstract class Analytics {
+  void close();
+
+  /// API to send events to Google Analytics to track usage
+  void sendEvent({
```
Should this be `Future<http.Response?> sendEvent(...)`? Also, whose responsibility should it be to check the status code of the response?

eliasyishak replied:
I originally left this as `void` because I didn't want the tools using this package to have to resolve any non-2xx status codes if the request was unsuccessful; instead it would silently fail

I think we may need a discussion what we should do here

christopherfujino replied:
You make a good point I hadn't considered that it may not be actionable :). On the other hand, why not return the response? The caller can always ignore it.

eliasyishak replied:
That's fair, including that in the next push

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -137,8 +194,64 @@ class AnalyticsImpl implements Analytics {
+  }
+
   @override
   void setTelemetry(bool reportingBool) {
```
I know it's not part of this PR, but I'm thinking `setTelemetry` should also be asynchronous, since we're both reading and writing to/from disk, which could be slow.

eliasyishak replied:
To make this method async, do I only make the `setTelemetry` method async and leave the other methods within `ConfigHandler` as they are (ie. keep i/o related methods like `readAsStringSync()`, `lastModifiedSync()`, etc.) or do those need to also change to their async counterparts?

Is it enough to do this?
```dart

/// NEW UPDATE
Future<void> setTelemetry(bool reportingBool) async {
  _configHandler.setTelemetry(reportingBool);
}
```

```dart

/// PREVIOUS
void setTelemetry(bool reportingBool) {
  _configHandler.setTelemetry(reportingBool);
}
```

christopherfujino replied:
you would need to make `_configHandler.setTelemetry` async, migrating readAsStringSync -> readAsString and writeAsStringSync -> writeAsString (I think lastModifiedSync is probably fast enough that it doesn't need to be async).

Also, did you mean to resolve this conversation?

eliasyishak replied:
Ah, I marked it as resolved because I thought I figured it out by wrapping the config handler's method to set telemetry status in a future.

If I switch the read and write operations to async, wouldn't I have to `await` for them anyway since I need the contents of the file to toggle telemetry status?

eliasyishak replied:
I have also refactored `setTelemetry` to the below code block
```dart
  @override
  Future<void> setTelemetry(bool reportingBool) {
    return Future<void>(() => _configHandler.setTelemetry(reportingBool));
  }
```

I tested it with the example file and it seems to be running asynchronously

christopherfujino replied:
> I have also refactored `setTelemetry` to the below code block
> 
> ```dart
>   @override
>   Future<void> setTelemetry(bool reportingBool) {
>     return Future<void>(() => _configHandler.setTelemetry(reportingBool));
>   }
> ```
> 
> I tested it with the example file and it seems to be running asynchronously

This is technically asynchronous, however, the slow parts (reading and writing from/to disk) are still synchronous. If you don't want your PR to be blocked by this, you can assign a TODO to me and I will refactor in another PR.

eliasyishak replied:
As discussed offline, the following issue https://github.com/eliasyishak/dash_analytics/issues/17 captures the work necessary for making this method async

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -41,8 +51,9 @@ abstract class Analytics {
     required int toolsMessageVersion,
     required String toolsMessage,
     required FileSystem fs,
```
Since this is a test helper, it might make sense to make some of these non-required and provide fallbacks. For instance, if the test doesn't need to verify what was written to disk, this factory could construct its own `MemoryFileSystem()`. Likewise for most (all?) the rest of these fields. This would make an easier API for writing tests, where you only need to provide the fields you want to test.

eliasyishak replied:
That was the initial design I had actually but I wanted to remove the default values so I know exactly what the test environment is using. For example, I need access to the file system from the tests to be able to access the files to check for their existence

```dart
fs = MemoryFileSystem.test();
home = fs.directory(homeDirName);

// The 3 files that should have been generated
clientIdFile = home
    .childDirectory(kDartToolDirectoryName)
    .childFile(kClientIdFileName);
sessionFile =
    home.childDirectory(kDartToolDirectoryName).childFile(kSessionFileName);
configFile =
    home.childDirectory(kDartToolDirectoryName).childFile(kConfigFileName);
});

// Checks below need access to the file system
test('Initializer properly sets up on first run', () {
  expect(clientIdFile.existsSync(), true,
      reason: 'The $kClientIdFileName file was not found');
  expect(sessionFile.existsSync(), true,
      reason: 'The $kSessionFileName file was not found');
  expect(configFile.existsSync(), true,
      reason: 'The $kConfigFileName was not found');
  expect(dartToolDirectory.listSync().length, equals(3),
      reason:
          'There should only be 3 files in the $kDartToolDirectoryName directory');
  expect(analytics.shouldShowMessage, true,
      reason: 'For the first run, analytics should default to being enabled');
});
```

christopherfujino replied:
By making them optional, tests that need to verify the file system can provide it, and the others don't need to provide it.

eliasyishak replied:
Makes sense, I've made them optional with default values for the message and version number, and if the filesystem isn't specified in the constructor, it will fall back on a default instance `MemoryFileSystem.test(style: ...)`

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -26,6 +32,10 @@ abstract class Analytics {
+        platform: platform,
+        toolsMessage: kToolsMessage,
+        toolsMessageVersion: kToolsMessageVersion,
+        fs: const LocalFileSystem(),
```
I don't think we should require callers to provide the `Directory homeDirectory` for two reasons:

1. we are relying on the clients to reproduce the logic to determine a user's home directory, which is complicated, and if two clients implement this differently, those two clients won't share the config file
1. the interface `Directory` is from package:file, and would require callers to do something like:

```dart
final homePath = determineUserHomePath();
final homeDir = (const LocalFileSystem()).file(homePath);
```

I think instead package:dash_analytics should take care of both of those steps.

It would make sense for the test class to have an optional override with a `Directory` object.

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -69,15 +81,35 @@ abstract class Analytics {
   late bool _showMessage;
+  late final GAClient _gaClient;
+  late final String _clientId;
+  late final UserProperty userProperty;
```
can this be private?

eliasyishak replied:
I left this one as not private because I thought it might be helpful for an IDE to have access to this possibly? It can provide visibility into what kind of user metadata we are sending. What do you think about exposing that kind of information?

christopherfujino replied:
SGTM

christopherfujino commented on test/dash_analytics_test.dart:
```diff
@@ -24,13 +29,16 @@ void main() {
+    home = fs.directory(homeDirName);
+    dartToolDirectory = home.childDirectory(kDartToolDirectoryName);
 
+    // The main analytics instance, other instances can be spawned within tests
```
I think two things are being conflated here:

1. client tests; and
2. package:dash_analytics' tests

This `TestAnalytics` class you've written makes sense for 1, however it is not sufficient for 2 since it shares no code with the real implementation. Essentially, `AnalyticsImpl` is untested.

It would make sense to have a few tests of `Analytics.test`, but essentially you should be testing `AnalyticsImpl` here.

eliasyishak replied:
I thought by extending the default `AnalyticsImpl` class for `TestAnalytics` and just overriding the `sendEvent` method I was testing the package. Everything else is shared between the two classes... should `TestAnalytics` instead be implementing the abstract class?

christopherfujino replied:
Oops, you're right, I read this wrong.

However, instead of overriding sendEvent, what about if in tests you instead use a `FakeGAClient` instance, that records the http requests it *would* have sent?

eliasyishak replied:
Good point, as discussed earlier today, I have captured this as an issue here https://github.com/eliasyishak/dash_analytics/issues/16 so that I can dedicate more time to working on mocking the client

christopherfujino commented on lib/src/config_handler.dart:
```diff
@@ -29,7 +31,7 @@ class ConfigHandler {
   /// yyyy-MM-dd (2023-01-09)
   static String get dateStamp {
-    return DateFormat('yyyy-MM-dd').format(DateTime.now());
+    return DateFormat('yyyy-MM-dd').format(clock.now());
```
Can you do something like:

```dart
class ConfigHandler {
  ConfigHandler({
    @visibleForTesting
    Clock? clockOverride,
    // etc
  }) : _clock = clockOverride ?? clock {
    // etc
  }
  
  final Clock _clock;
}
```

This way test-owned clocks can be provided.

eliasyishak replied:
This was initially how I did this but realized that this would required having to pass the clock from the top level `Analytics` class to every class it initializes.

Instead I read that `package:clock` avoids this for testing by using the `withClock(Clock clock, T Function() callback)` method. This allows us to pass a fake clock as the first positional parameter and it will essentially do what you have outlined but through the use of zones. I referred to the article below

https://iiro.dev/controlling-time-with-package-clock/

christopherfujino replied:
I'll leave this one up to you as I don't think there are too many tests clients would write with a fake clock, but I'll point out that you're direct injecting everything else, and using `withClock()` is essentially the opposite direction of https://github.com/flutter/flutter/issues/47161

christopherfujino commented on lib/src/config_handler.dart:
```diff
@@ -131,7 +132,7 @@ class ConfigHandler {
 
     // Update the [ToolInfo] object for the current tool
-    parsedTools[tool]!.lastRun = DateTime.now();
+    parsedTools[tool]!.lastRun = clock.now();
```
You can avoid the `!` operator if you rewrite the first statement in this function as:

```dart
final ToolInfo? toolInfo = parsedTools[tool];
if (toolInfo == null) {
  return;
}
```

And then these lines can be (via type promotion):

```dart
    // Update the [ToolInfo] object for the current tool
    toolInfo.lastRun = clock.now();
    toolInfo.versionNumber = newVersionNumber;
```

christopherfujino commented on test/dash_analytics_test.dart:
```diff
@@ -1,10 +1,14 @@
+import 'dart:convert';
```
This (and every other file in this repo) needs to have a LICENSE comment. This also needs a top level LICENSE file.

christopherfujino commented on lib/src/constants.dart:
```diff
@@ -1,3 +1,13 @@
+const String kClientIdFileName = 'CLIENT_ID';
+
+/// Name for the file where telemetry status and tools data will be stored
+const String kConfigFileName = 'dart-flutter-telemetry.config';
```
can this be `dash-analytics.config`, so it's obvious to someone doing `ls .dart-tool` what it is?

christopherfujino commented on lib/src/constants.dart:
```diff
@@ -37,9 +47,16 @@ reporting=1
 const int kSessionDurationMinutes = 30;
 
+/// Name for the json file where the session details will be stored
+const String kSessionFileName = 'session.json';
```
can we namespace this, like `dash-analytics-session.json`, or something similar?

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -0,0 +1,28 @@
+  ;
+
+  final String label;
+  final String desc;
```
nit
```suggestion
  final String description;
```

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -0,0 +1,28 @@
+    desc: 'Hot reload duration',
+  ),
+  ;
+
```
How about we add:

```dart
final String toolOwner;
```

Which, for `hot_reload_time` would be `flutter_tools`.

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -0,0 +1,28 @@
+}
+
+/// Enumerate options for platform
+enum DevicePlatforms {
```
```suggestion
enum DevicePlatform {
```

christopherfujino replied:
Where and for what purpose is this used? Are we not currently sending more detailed operating system information with each event?

eliasyishak replied:
This is being used by the tools implementing this package. The idea was to have each client tool determine what their OS is, then match it against one of these enum values. Once they have assigned an enum value, we will have all relevant information around the description, tool owner, and anything else we want to associate with OS's

It is better than having each client send a string representation of the OS which could vary (ie. macos, macOS, etc.)

christopherfujino replied:
Two things:

1. Why not have this package detect the operating system?
2. Ahh, I thought that we sent the OS version with each event, but I was wrong.

eliasyishak replied:
Good idea, I've moved the logic into the package now

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -0,0 +1,28 @@
+/// Enumerate options for platform
+enum DevicePlatforms {
+  windows('Windows'),
+  macos('Macintosh'),
```
I think this should be the name of the operating system:

```suggestion
  macos('macOS'),
```

Unless there is some other constraining factor?

eliasyishak replied:
I'm indifferent to what we call them, I just asked JaYoung what they called them currently in GA3 and continued. I can relabel them though

christopherfujino commented on lib/src/ga_client.dart:
```diff
@@ -0,0 +1,53 @@
+    required this.apiSecret,
+  })  : postUrl =
+            '$kAnalyticsUrl?measurement_id=$measurementId&api_secret=$apiSecret',
+        _client = http.Client();
```
should we allow tests to provide their own (fake) client?

eliasyishak replied:
This could be useful, but I'm not sure how helpful it will be because the GA servers always send 2xx status codes, even if the data is malformed. Additionally, the server doesn't send back any payload. Open to discuss this further

christopherfujino replied:
Ahh, I didn't realize they always return 2xx, lol. I think there's still value in providing a fake client as it means in tests you can verify the request looks correct.

christopherfujino commented on lib/src/initializer.dart:
```diff
@@ -69,20 +33,59 @@ class Initializer {
+  /// yyyy-MM-dd (2023-01-09)
   String get dateStamp {
-    return DateFormat('yyyy-MM-dd').format(DateTime.now());
+    return DateFormat('yyyy-MM-dd').format(clock.now());
```
As with `config_handler.dart`, can we allow for direct injecting test-controlled clocks?

christopherfujino commented on lib/src/session.dart:
```diff
@@ -0,0 +1,84 @@
+  final Directory homeDirectory;
+  final FileSystem fs;
+  final File _sessionFile;
+  late int _sessionId;
```
avoid `late` except in tests when they are initialized by `setUp()`. One trick to allow a field to be non-nullable when it has to be calculated from a function call is to create a factory that makes the necessary calls (here calls to `.readAsStringSync()` and `jsonDecode`, then returns the result of a call to a private constructor that then initializes the non-nullable fields (e.g. `Session._({this._sessionId, this._lastPing, etc...})`).

eliasyishak replied:
This was a great idea, thank you!

christopherfujino commented on lib/src/session.dart:
```diff
@@ -0,0 +1,84 @@
+  /// because the last ping variable will always need to be persisted
+  int getSessionId() {
+    _refreshSessionData();
+    final DateTime now = clock.now();
```
Again, consider allowing tests to provide their own clock.

eliasyishak committed:
update to analysis options + using specific typing for `userPropertyMap`

eliasyishak committed:
adding necessary types after enabling lint

eliasyishak committed:
adding documentation for `TestAnalytics` class

eliasyishak committed:
checking for tool in parsedTools map

eliasyishak committed:
various nits

- `DevicePlatform` instead of `DevicePlatforms`
- Renaming files for config and session.json
- Relabeling desc to description in DashEvents
- Adding toolOwner variable

eliasyishak committed:
cleanup

eliasyishak committed:
refactoring session.dart to use factory constructor to avoid late keyword

eliasyishak committed:
need to extract the label from the enum

eliasyishak commented on example/dash_analytics_example.dart:
```diff
@@ -35,18 +33,23 @@ final Analytics analytics = Analytics(
   tool: tool,
   homeDirectory: getHomeDirectory(),
```
@christopherfujino do you suggest that this method that resolves the user's home directory be moved into the package?

christopherfujino replied:
I think that would be a good idea

eliasyishak committed:
relabeling directory

eliasyishak committed:
adding license info to all dart files + LICENSE file

eliasyishak committed:
updating example of payload

eliasyishak committed:
resolving home directory for default implementation of analytics

eliasyishak committed:
returning future from sendEvent method to caller

eliasyishak committed:
Update dash_analytics_test.dart

eliasyishak committed:
check for os for test suite

eliasyishak committed:
wrong year in license top line comment

eliasyishak committed:
converting setTelemetry to be async

eliasyishak committed:
utility function created to generate request body

eliasyishak committed:
fixing tests to work with setTelemetry's async process

eliasyishak committed:
marking message, version, and fs as optional for test analytics constructor

eliasyishak committed:
test added to check for malformed request bodys

eliasyishak committed:
adding async keywords where necessary

eliasyishak committed:
reverting async + adding TODO for separate PR

christopherfujino reviewed (approved):
LGTM once there's a TODO about setTelemetry
